src/rdkafka_topic.c

```diff
@@ -924,36 +926,73 @@ static void rd_kafka_topic_assign_uas (rd_kafka_topic_t *rkt,
 
 
 /**
- * Received metadata request contained no information about topic 'rkt'
- * and thus indicates the topic is not available in the cluster.
+ * @brief Mark topic as non-existent, unless metadata propagation configuration
+ *        disallows it.
+ *
+ * @param err Propagate non-existent topic using this error code.
+ *            If \p err is RD_KAFKA_RESP_ERR_TOPIC_EXCEPTION it means the
+ *            topic is invalid and no propagation delay will be used.
+ *
+ * @locks topic_wrlock() MUST be held.
  */
-void rd_kafka_topic_metadata_none (rd_kafka_topic_t *rkt) {
-	rd_kafka_topic_wrlock(rkt);
+static rd_bool_t rd_kafka_topic_set_notexists (rd_kafka_topic_t *rkt,
+                                               rd_kafka_resp_err_t err) {
+        rd_ts_t remains_us;
+        rd_bool_t permanent = err == RD_KAFKA_RESP_ERR_TOPIC_EXCEPTION;
+
+        if (unlikely(rd_kafka_terminating(rkt->rkt_rk))) {
+                /* Dont update metadata while terminating. */
+                return rd_false;
+        }
 
-	if (unlikely(rd_kafka_terminating(rkt->rkt_rk))) {
-		/* Dont update metadata while terminating, do this
-		 * after acquiring lock for proper synchronisation */
-		rd_kafka_topic_wrunlock(rkt);
-		return;
-	}
+        remains_us =
+                (rkt->rkt_ts_create +
+                 (rkt->rkt_rk->rk_conf.metadata_propagation_max_ms * 1000)) -
+                rkt->rkt_ts_metadata;
+
+        if (!permanent &&
+            rkt->rkt_state == RD_KAFKA_TOPIC_S_UNKNOWN && remains_us > 0) {
+                /* Still allowing topic metadata to propagate. */
+                rd_kafka_dbg(rkt->rkt_rk, TOPIC|RD_KAFKA_DBG_METADATA,
+                             "TOPICPROP",
+                             "Topic %.*s does not exist, allowing %dms "
+                             "for metadata propagation before marking topic "
+                             "as non-existent",
+                             RD_KAFKAP_STR_PR(rkt->rkt_topic),
+                             (int)(remains_us / 1000));
+                return rd_false;
+        }
 
-	rkt->rkt_ts_metadata = rd_clock();
+        if (!err)
+                err = RD_KAFKA_RESP_ERR__UNKNOWN_TOPIC;
```

Comment:
this seems like it should be a pre-condition check (assert). a bit dodgy to have it as an allowed path?